Improve memory usage for activations

mciprian13 opened the pull request:
**Summary**
Improve memory allocator by adding an `allocateAll` function which has visibility to all the allocations at once and hence can do a better allocation (with higher efficiency) for activations. In principle this is done by allocating the larger buffers first and thus avoiding early fragmentation (BestFit).

**Test Plan**
Added unit tests in MemoryAllocator containing:
- feature testing
- negative testing
- activation allocation for 79 models (the unit tests were automatically generated using a Matlab script using the activation pattern extracted from the IR dumped by Glow)

mciprian13 commented:
@opti-mix Interested in reviewing this optimization?

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  return usedSizeMax;
+}
+
+uint64_t MemoryAllocator::allocateAll(const std::list<Allocation> &allocList) {
```
This function is rather complex. Maybe split it into a main function and a bunch of helpers?

mciprian13 replied:
Ok, I`ll look into this.

mciprian13 replied:
I removed the part which maps handles to IDs and put it in a helper.
In order to make also the part which gathers buffer information as a separate utility I would need to make all the other definitions part of the MemoryAllocator class and that I think would pollute the header file with all sorts of stuff which are only specific to the implementation and not the API. WDYT?

opti-mix replied:
Hmm. 

> I think would pollute the header file with all sorts of stuff which are only specific to the implementation and not the API.

Could it help if you add something like this to the MemoryAllocator.h:

```c++
// Forward declaration of MemoryAllocatorImpl.
class MemoryAllocatorImpl.
...
class MemoryAllocator {
friend class MemoryAllocator;
...
};
```

So, basically you say there is friend class that can access all the info from MemoryAllocator. And then you define the actual implementation of a friend class MemoryAllocatorImpl in the cpp file.

WDYT? Would it work?

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  }
+
+  // ---------------------------------------------------------------------------
+  // Allocate all the buffers using all the available strategies.
```
How does it affect the compilation time if you try to allocate using all available strategies?

mciprian13 replied:
The function tries to use all the strategies but does not necessarily use them all. If the first strategy obtains the optimal allocation (with maximum efficiency) then the others are not used anymore. This is the case for most of the models (I think 70 out of the 79 models in the unit tests) the maximum efficiency is obtained with the first strategy.
Regarding compile-time the algorithm runs in a couple of ms.
e.g. The maximum it takes for the Model67 (resnet_152) which has lots of layers and for my VM it takes around 17 ms (the compile-time itself is in the order of 10's of seconds so this should be insignificant).

opti-mix commented on tests/unittests/MemoryAllocatorTest.cpp:
```diff
@@ -342,3 +342,9199 @@ TEST(MemAlloc, testAlignment) {
+#endif
+}
+
+/// Test memory allocation for model cifar10_quant.tflite.
```
If I understand this code correctly you recorded the memory allocation requests for a number of models and try to check how they are handled. I'm a bit concerned that we hard-code those numbers here. Loading a model and doing memory allocation would be more reliable....

opti-mix replied:
Also, most of the code in this file is auto-generated I assume. And the file is huge. Could you instead dump the allocs per model into an input file and then in this CPP file just read them, create a sequence of allocations and use the allocator then? This would make this source file MUCH smaller.

mciprian13 replied:
Indeed I recorded the allocations from 79 models (by parsing their corresponding IR) and auto-generated unit tests here using a Matlab script. I don't see why hardcoding would be a problem since the unit tests implicitly verify the allocation is correct (segments are always validated within the function `allocateAll`) and also have a regression of the allocation efficiency (you can verify that for each unit tests the efficiency is checked against the reference) which is more than enough to test the allocator.
Compiling all the models directly would be an overkill since that means all the models should be downloaded during testing (this means 2.36 GB).
I agree we could compress the unit tests. We have 2 options:
1) either keep only a couple of unit tests for models (say 10)
2) we could export the information into files but in order to save space we should export them as binaries but this makes the tests harder to read and requires some extra code for parsing. Any suggestions?

opti-mix replied:
Yes, compiling all models would be an overkill. Let's compress the unit tests.

> we could export the information into files but in order to save space we should export them as binaries but this makes the tests harder to read and requires some extra code for parsing. Any suggestions?

You could use text files with the recorded allocs. The idea is to keep this information that is purely data and semantically is not very important out of sight when you look at the test. The test should be small and simple:
- load data from a file
- build a list of allocs
- run the allocator
- check the results/quality of allocations

mciprian13 replied:
Should I make one text file for each model (e.g. `Model1.txt`, `Model2.txt`, etc)? Where to put these files: in a separate folder inside the unit tests is ok, e.g. `MemoryAllocatorTestModels`?
Or a single file called `MemoryAllocatorTest.txt` in the same folder?

opti-mix replied:
> in a separate folder inside the unit tests is ok, e.g. MemoryAllocatorTestModels?

Yes, this would be the better solution.

mciprian13 replied:
This is done!

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+/// The strategies are listed in the decreasing order of the likelihood of
+/// getting the best allocation efficiency. These strategies are only used
+/// within the allocateAll function used to allocate all segments at once.
+enum class MemAllocStrategy : uint8_t {
```
Have you considered allowing for custom memory allocation strategies? I.e. there are some supported by default, but users could register their own with a given instance of an allocator. Something like:

```c++
MemoryAllocator::MemoryAllocator(...) {
   /// Register standard strategies.
   registerMemAllocStrategy(MaxLiveSizeMaxBuffSizeImplClosure);
   ...
}
/// And then optionally later in user-code.
MemoryAllocator myAlloc(...);
/// Register my own mem alloc strategy implementation.
myAlloc.registerMemAllocStrategy(customMemAllocStrategyImplClosure);
```

mciprian13 replied:
It is an interesting idea but I don't think this is necessary. The goal of the memory allocator is provide allocation with best efficiency (lowest memory usage). I don't think users would be interested in the details and in registering their own strategies.

opti-mix replied:
The users of allocators are e.g. Glow backends. Some backends may have very target specific allocation strategies, which should not be part of the general MemoryAllocator.

But I agree that it is a pretty advanced scenario and is probably not very important.

mciprian13 replied:
For someone to register its own strategy would require some additional context (e.g. the mapping between segment IDs and handles). Since I don't have yet an example of how a user would want to specialize the allocation, I think it is best to leave this feature for later.

opti-mix reviewed (commented):
@mciprian13 Thanks for this contribution. This is an interesting generalization!

But it adds a lot of code and many functions are pretty long and complex. It would be nice to make them smaller and easier to understand. I made some suggestion in my review comments.

Also, you actively use std::list in many places even though you don't remove from the middle of the sequence or something like that. Is there any reason for using list instead of a vector in such cases?

Another quick question: You extended the MemoryAllocator class directly. Maybe different strategies or extended functionality should be implemented by deriving from it?

mciprian13 commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -119,35 +170,64 @@ class MemoryAllocator {
-  uint64_t maxMemoryAllocated_{0};
+
+  /// A list of allocated segments.
+  std::list<Segment> segments_;
```
I just renamed the variable from `allocations_` to `segments_`. The variable was `std::list` from the beginning because the functions `allocate`/`deallocate` introduce/erase segments dynamically.

opti-mix replied:
I see. What about other places where you use std::list? You seem to be using it very actively in this PR.

mciprian13 replied:
I use list for containers which grow dynamically with `push_back` and it's not easy to know before hand it's size in order to reserve memory (for example the liveness information). The input allocations are also a list because of the same reason: the way it's constructed in `AllocationsInfo.cpp`.

opti-mix replied:
Yes, you don't know the exact size in advance, but you know the upper bound, e.g. the number of IR instructions, so you can just reserve it in the container from the beginning and it will be only one dynamic allocation. std::list created new dynamic allocations for each element you add. So, unless you need to add/remove elements in the middle, there is no good reason for lists.

But this is a relatively minor thing.

mciprian13 replied:
The only place I use now std::list is:
- when I have list of live IDs in the algorithm which are consumed randomly (elements removed from anywhere)
- the front-end allocation list provided by the user.
Do you want still to change the allocation list to `std::vector`?

mciprian13 commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+/// by the function \ref allocateAll about the buffer sizes and live intervals.
+static uint64_t allocateAllWithStrategy(
+    size_t buffNum, size_t allocNum, uint64_t memorySize,
+    const std::vector<uint64_t> &buffSizeArray,
```
Ok, I`ll do this.

mciprian13 replied:
I created two additional structures `BuffInfo` and `LiveInfo` which merge 2 or more such arrays.
Done!

mciprian13 commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  for (size_t buffIdx = 0; buffIdx < buffNum; buffIdx++) {
+
+    uint64_t currSegId;
+    if (strategy == MemAllocStrategy::MaxLiveSizeMaxBuffSize) {
```
Ok I`ll do this.

mciprian13 replied:
I created a separate function for each strategy (with same API) and populated a vector of functions. As a result this has been moved outside the function.
Done!

mciprian13 commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  // Allocate buffers in the same order as they were requested.
+  SameOrder,
+  // Number of allocation strategies. Should remain the last enumeration value.
+  Count
```
Not relevant anymore since now I have a vector of functions.

mciprian13 commented:
@opti-mix I refactored the large parts you requested. There are some minor things to be dealt with (see the comments). Can you have another look? Thanks!

opti-mix reviewed (commented):
Looks much better now, thanks! Please see my comments.

mciprian13 commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,580 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+/// strategies are used one after the other in the exact order as listed here.
+/// In case one such strategy provides the maximum efficiency (best theoretical
+/// result) then the following ones (if any) are not used anymore.
+static std::vector<MemAllocStrategy> MemAllocStrategyArray = {
```
Renamed it to `memAllocStrategies`.
Done!

mciprian13 commented on tests/unittests/MemoryAllocatorTest.cpp:
```diff
@@ -342,3 +344,221 @@ TEST(MemAlloc, testAlignment) {
+
+/// Test memory allocation for all models.
+TEST(MemAlloc, testAllocateAllForModels) {
+  for (int modelIdx = 1; modelIdx <= 79; ++modelIdx) {
```
- I added a print which tells which model is currently allocating. When running the test it looks like this:
![image](https://user-images.githubusercontent.com/29785908/102215494-761bda80-3ee2-11eb-8e20-b2999f31443d.png)
- I merged all the text files into one called `MemoryAllocatorTextModels.txt` which is now in the same directory as the unit test itself `MemoryAllocatorTest.cpp`
Done!

mciprian13 commented:
@opti-mix Having a friend class would require some methods to be provided with a reference to the MemoryAllocator object and would also break encapsulation. I don't know if just for the purpose of breaking down a function we really need such a major refactoring.

opti-mix commented:
@mciprian13 

> Having a friend class would require some methods to be provided with a reference to the MemoryAllocator object and would also break encapsulation. I don't know if just for the purpose of breaking down a function we really need such a major refactoring.

Yeah, I agree. Let's leave this idea for now.

opti-mix commented:
@mciprian13 I imported this PR into our internal build system. Here are the two issues that popped up:

1) Nit: static analyzers suggest to use emplace_back instead of push_back in many places.
2) The test fails, because it tried to open the file in the current directory, but I guess it is not guaranteed that the test is always run in the same directory where the source file lives. We need to find a way how to open reliably under any circumstances.

```
MemoryAllocatorTest: glow/tests/unittests/MemoryAllocatorTest.cpp:535: virtual void MemAlloc_testAllocateAllForModels_Test::TestBody(): Assertion `fs.is_open() && "Error opening file!"' failed.
```

mciprian13 commented:
> @mciprian13 I imported this PR into our internal build system. Here are the two issues that popped up:
> 
> 1. Nit: static analyzers suggest to use emplace_back instead of push_back in many places.
> 2. The test fails, because it tried to open the file in the current directory, but I guess it is not guaranteed that the test is always run in the same directory where the source file lives. We need to find a way how to open reliably under any circumstances.
> 
> ```
> MemoryAllocatorTest: glow/tests/unittests/MemoryAllocatorTest.cpp:535: virtual void MemAlloc_testAllocateAllForModels_Test::TestBody(): Assertion `fs.is_open() && "Error opening file!"' failed.
> ```

1. You want me to have another look?
2. For the unit test I also added a CMake rule to copy the file "MemoryAllocatorTestModels.txt" in the binary directory where the unit test is built. I guess for your build system you need that also. But even with this the unit test requires to be run from a specific path.

opti-mix commented:
> 1. Can you provide some file/line numbers to look at?

Yes, it would be nice if you'd replace almost all push_back occurrences by emplace_back.

As for (2), let me see what could be done.

opti-mix commented:
@mciprian13 

Regarding the issue with accessing a file. @jfix71 explained that we have a mechanism for this already. Search for GLOW_DATA_PATH e.g. in OnnxImporterTest.cpp -- we define GLOW_DATA_PATH in GLOW_DEFS and build the test with it, which prepends the correct path locally. For CMake based builds this GLOW_DATA_PATH will be just empty, but for our internal builds it will be initialized properly and form the right path. So, just try to follow the same pattern.

mciprian13 commented:
@opti-mix
1. I replaced the `push_back` with `emplace_back` in all the places
2. I added `GLOW_DATA_PATH` to the path. It is not clear to me whether this variable points to the Glow source code or the Glow build folder. For the CMake based build the text file is copied in the Glow build folder and used from there. For your build system it might be the source code folder, if not the text file should be copied as well. Please have a try!

opti-mix commented:
@mciprian13 Thanks for the fixes! It fixed the issue with the unit test.

But we have a different problem now on one of the backends not directly related to this PR, except that they use the MemoryAllocator. 

Have you changed any default behaviors of MemoryAllocator in this PR? I.e. would the old clients of the MemoryAllocator see any change in the behavior?

opti-mix commented:
@mciprian13 I think the issue we observe has something to do with the following:

- You changed `AllocationsInfo::allocateActivations` to do use allocateAll by default:
```
  // Allocate all segments at once for better allocation efficiency.
  activationsAllocator_.allocateAll(allocList);
```

- Some of our backends fail with it. Most likely it is because some of them reserve some space at the start of the activations memory block for its needs by calling `activationsAllocator.allocate(fixed_reserved_space, handle);`. This space is not supposed to be used for the allocation of any activations. All of them start after this reserved space in the beginning. Your code in `allocateAll` seems to reset() the allocator though and thus probably breaks the assumptions of those backends.

- Also there is a method called `allocateActivations` in `BackendUtils.cpp`. This method so far repeated the logic in `AllocationsInfo::allocateActivations`, but with your change they would diverge which may lead to issues. These methods should be kept in sync. Could you please make sure it is the case? In principle, both methods should reuse the same implementation and not duplicate the logic, but it is not done yet unfortunately. There is an issue for that: #3011.

I see different ways forward here:
- Your PR may leave the default behavior of `AllocationsInfo::allocateActivations` as it was and add some knob to change it to use `allocateAll`. This would be the preferred short-term solution if it works for you.
- We need to change something on our side to make it work with your PR. It can be done, but may take longer ;-)

I'll try to fix it quickly on our side, but it may not work out.

opti-mix commented:
@mciprian13 

> Your PR may leave the default behavior of AllocationsInfo::allocateActivations as it was and add some knob to change it to use allocateAll. This would be the preferred short-term solution if it works for you.

I can confirm that leaving the old logic in `AllocationsInfo::allocateActivations` as a default not break any other backends.

> I'll try to fix it quickly on our side, but it may not work out.

I checked and there is no quick fix.

mciprian13 commented:
@opti-mix The main reason why I did this PR is to replace the default allocation of activations with the new method which is significantly more efficient. The method used by your backends to allocate some space for their needs seems very hackish to me. They could instead implement their own `allocateActivations` function to do that. Alternatively we could add a flag to `allocateActivations` to specify what allocation method to use (but the default should still remain the one based on `allocateAll`).

When is the `allocateActivations` from `BackendUtils.cpp` used and why does it have to be sync'ed with `AllocationsInfo::allocateActivations`? I`m asking this because it's not like the code is duplicated because the code looks pretty different.

opti-mix commented:
> The main reason why I did this PR is to replace the default allocation of activations with the new method which is significantly more efficient. 

Yes, it is more efficient, no doubts. But memory consumption is not a very big concern for non memory-constrained targets. This is why it was not addressed so far.

> The method used by your backends to allocate some space for their needs seems very hackish to me. 

That's true. We'll see if we can fix it. 

> They could instead implement their own allocateActivations function to do that. 

True. But it will take some time. We cannot land this PR and just break them. We need a workaround.

> Alternatively we could add a flag to allocateActivations to specify what allocation method to use (but the default should still remain the one based on allocateAll).

I think this would be the easiest short-term solution. Once we migrated our code to work correctly with allocateAll, we could remove this temporary solution. Could you add such a flag and APIs to set/get it?

> When is the allocateActivations from BackendUtils.cpp used and why does it have to be sync'ed with AllocationsInfo::allocateActivations? I'm asking this because it's not like the code is duplicated because the code looks pretty different.

allocateActivations from BackendUtils.cpp is invoked from `runtime::RuntimeBundle::create`, which in turn is invoked from `Interpreter::compileIRWithoutConstants` and from `LLVMBackend::compileIRWithoutConstants`.

opti-mix commented:
> When is the allocateActivations from BackendUtils.cpp used and why does it have to be sync'ed with AllocationsInfo::allocateActivations? I`m asking this because it's not like the code is duplicated because the code looks pretty different.

Also, even though the allocateActivations logic is not literally the same (in terms of code), it was assigning the same offsets to the same activations so far. This invariant should be better kept in-sync. If not, it may break the correctness of AOT-compilation or JIT-compilation I think.

opti-mix commented:
> Your code in allocateAll seems to reset() the allocator though and thus probably breaks the assumptions of those backends.

Related to this: Please have a look e.g. at the BundleSaver::performBundleMemoryAllocation(). BundleSaver may be used to emit multiple IR functions into the same bundle. Thus, it reuses the same AllocationsInfo and does not reset it every time it calls any methods from it. As a result, the same set of allocators will be used to allocate constant weights, mutable weights and activations. The net effect will be that it would allocate for each region the cumulative amount of memory required by all the IR functions being saved into a bundle.

Looks like there is no unit-test for this yet though ;-)

mciprian13 commented:
ping

opti-mix commented:
@mciprian13 Please address my two last comments in this thread about  allocateActivations from BackendUtils.cpp being in sync with AllocationsInfo::allocateActivations. Your PR should not break invariants, at least not in the default mode of operation.

mciprian13 commented:
@opti-mix I will have a look on your comments this week.

mciprian13 commented:
@opti-mix
1. I had a look over `BackendUtils.cpp` but in my view even before this PR the memory allocation there was different than the one used in `AllocationsInfo::allocateActivations` and that is because in `BackendUtils.cpp` the function `allocateActivations()` also handles the TensorViews but more importantly has an extra CLI parameter `reuseActivationsMemory` which adds some more functionality than `AllocationsInfo::allocateActivations` (which seems to operate always under the assumption that reuseActivationsMemory = true because activations are always freed when going from one function to the next). I don't know how to proceed for sync'ing the two functions because they look pretty different.

2. Indeed in BundleSaver there might be multiple IR functions and currently the logic I implemented will break the functionality because the allocator is always cleared when allocating for a certain function. I can modify the function `allocateAll` by:
    - not clearing the memory allocator to enable to call this function multiple times with same allocator but different IR functions
    - adding a bool flag called `reuseMemory` such that e.g. when allocating activations for Nth function to specify whether we reuse or not the memory allocated for the previous functions

The problem with 2) is that I don't know how to write an end-to-end test with a bundle with multiple entries executed with CPU backend because ONNX/TFLite formats do not allow defining models with multiple graphs (functions).

opti-mix commented:
@mciprian13 

> I had a look over BackendUtils.cpp but in my view even before this PR the memory allocation there was different than the one used in AllocationsInfo::allocateActivations and that is because in BackendUtils.cpp the function allocateActivations() also handles the TensorViews but more importantly has an extra CLI parameter reuseActivationsMemory which adds some more functionality than AllocationsInfo::allocateActivations (which seems to operate always under the assumption that reuseActivationsMemory = true because activations are always freed when going from one function to the next). I don't know how to proceed for sync'ing the two functions because they look pretty different.

IIRC, this reuseActivationsMemory flag was introduced for debugging purposes only, to make sure each intermediate result is updated only once and you can easily inspect it later.  And it was almost never actively used. So, this flag is not very important. Even if you break it - it is fine ;-)

Regarding (2):
> - not clearing the memory allocator to enable to call this function multiple times with same allocator but different IR functions
> - adding a bool flag called reuseMemory such that e.g. when allocating activations for Nth function to specify whether we reuse or not the memory allocated for the previous functions
> The problem with 2) is that I don't know how to write an end-to-end test with a bundle with multiple entries executed with CPU backend because ONNX/TFLite formats do not allow defining models with multiple graphs (functions).

I could think of multiple different ways to create such a test:
- Manually created test, where you create two simple functions in the same module.
- Using Glow partitioner with artificially configured low memory thresholds, which would split a loaded model into smaller partitions and thus produce multiple functions.

mciprian13 commented:
@opti-mix I fixed the last 2 things:
1) I modified the `allocateAll` to be capable to allocate for multiple functions without resetting the allocator. I added some extra unit tests in `MemoryAllocatorTest`. I also found that there already was an unit test called `bundle_with_multiple_entries` which was not used. I modified it a little bit to make sure the two IR functions use the activations memory by using only MatMul nodes (the previous one had only in-place operations and did not use the activations memory). I added this test to the `RELEASE_WITH_EXPENSIVE_TESTS` CI build.
2) I modified the `BackendUtils.cpp` source file to mirror the logic from `AllocationsInfo.cpp`. I removed the CLI flag `reuseActivationsMemory` since I couldn't use it anymore with the `allocateAll` function (the logic behaves now as if it is always true).
Hopefully this was all that needed to be done. Have another look!

mciprian13 commented:
@opti-mix After modifyin the `BackendUtils.cpp` it seems that the OpenCL unit tests are failing. Do you know why?

mciprian13 commented:
@opti-mix It seems there is more to the story behind `BackendUtils.cpp`. Now looking at it it does not resemble at all the logic from   `AllocationsInfo` because in `runtime::RuntimeBundle::create` all the graph operands (placeholders, constants, activations) are allocated using the same allocator object (in AllocationInfo we use different memory allocators). Currently the function `allocateAll` is not intended to be mixed with the existing function `allocate()`. This means the modification in `BackendUtils.cpp` won't work. This raises even more doubts why this logic should be kept in sync with AllocationInfo because IMO it was never in sync. WDYT?

opti-mix commented:
> It seems there is more to the story behind BackendUtils.cpp. Now looking at it it does not resemble at all the logic from AllocationsInfo because in runtime::RuntimeBundle::create all the graph operands (placeholders, constants, activations) are allocated using the same allocator object (in AllocationInfo we use different memory allocators).

Yeah. There are 3 different `runtime::RuntimeBundle::create` methods. Some of them are used by the LLVMBackend-derived classes and some others are not.

The overall idea of allocating memory for constants, placeholders and activations currently is:
- Constants/Constant weights do not reuse memory of each other or any other kinds of allocated objects.
- Placeholders do not reuse memory of each other or any other kinds of allocated objects.
- Activations can reuse memory of each other if their lifetimes do not overlap.

These rules are currently true even if the same allocator object is used for all of them and even more so if different allocator objects are used.

Note: *The reason why we do not mix/reuse addresses for different kinds of symbols is that we consider them to be logically in different "address spaces". For some targets it can literally be the case. E.g. constants may be stored in a read only/constants memory on GPUs or some custom accelerators. Placeholders may also be stored in dedicated memories which are used for e.g. DMA transfers to/from a host. And activations may leave e.g. in DRAMs/SRAMs. Generally, the decision about what kind of things or what things can or cannot be reused is very target dependent. In the future, we may extend the implementations to allow backends specify their custom memory allocation policies. But this goes beyond the scope of this PR.*

Also a lot of logic in `AllocationsInfo` and `BackendUtils.cpp` is duplicated. I'll try to submit a PR where the address assignment logic in `AllocationsInfo` will be literally calling the same implementation that is used by the RuntimeBundle. This way we ensure that the address assignment is the same and is implemented in one place only. Your PR would then need to update this single place only.


> Currently the function allocateAll is not intended to be mixed with the existing function allocate(). This means the modification in BackendUtils.cpp won't work. 

I think the only purpose of allocateAll is to be used allocate activations, right? Thus, memory allocated before this call (e.g. for constants or placeholders) can be considered as untouchable. I.e. if upon entering allocateAll the current top address of the memory allocator is TopAddr, any addresses assigned by allocateAll should be >= TopAddr. WDYT?

mciprian13 commented:
@opti-mix Yes the function `allocateAll` is intended to be used for activations only (that's the only place where a fancier algorithm is needed). But your proposal concerns me: do you propose to use the same allocator object for all the operand types (placeholders, constants, activations)? If yes, doesn't this break the current design of memory for the bundles where each operand type uses a separate allocator object? Hopefully we don't change the memory allocation for bundles because we have other private features depending on it.

opti-mix commented:
>  But your proposal concerns me: do you propose to use the same allocator object for all the operand types (placeholders, constants, activations)? If yes, doesn't this break the current design of memory for the bundles where each operand type uses a separate allocator object? Hopefully we don't change the memory allocation for bundles because we have other private features depending on it.

Don't worry. I do not propose to use the same allocator object for all the operand types. I'm saying that backends can control if they want to use different or same allocators for some of the operand types. In particular, all LLVMBackend-derived backends (who are the only ones producing bundles today) currently use different allocators for each operand type and I'm not proposing to change this.

opti-mix commented:
@mciprian13 

> Also a lot of logic in AllocationsInfo and BackendUtils.cpp is duplicated. I'll try to submit a PR where the address assignment logic in AllocationsInfo will be literally calling the same implementation that is used by the RuntimeBundle. This way we ensure that the address assignment is the same and is implemented in one place only. Your PR would then need to update this single place only.

Have a look at #5301

mciprian13 commented:
@opti-mix Although your PR solves the code sync problem, my algorithm cannot be used where same allocator is used with `allocate()` (e.g. used to allocate constants) and 'allocateAll()` (e.g. used to allocate activations). One such backend which uses same allocator for all the operand types is the OpenCL backend. As you can see for latest tests, using `allocateAll` for activations breaks the OpenCL backend. This problem seems harder to fix. One idea is to not use my algorithm at all and only those backends who want to use it to do so.

opti-mix commented:
@mciprian13 

> my algorithm cannot be used where same allocator is used with allocate() (e.g. used to allocate constants) and 'allocateAll()`

OK. If you cannot easily re-factor your algorithm to ignore all allocations before the call to allocateAll, there is possibly another workaround:
- What if you use a new allocator NewAlloc for allocateAll and you use it to allocate all activations and assign addresses to them.
- Once this is done, you use these addresses as offsets relative to the top address of the actual allocator that has been used to allocate something before the call to allocateAll. I.e. if the top address of the actual allocator was TopAddress and the address of an activation A assigned by NewAlloc is ActAddress, then the address of A in the symbol table or in the oldAlloc should become TopAddress + ActAddress.

I hope I managed to explain the general idea. Do you think something like this may work?

mciprian13 commented:
@opti-mix I will rebase on your latest commit and see what I can do.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -119,35 +175,70 @@ class MemoryAllocator {
+  /// \p size (in bytes). \returns the new memory usage or MemoryAllocator::npos
+  /// if the new memory usage exceeds the memory size. The given \p size when
+  /// used internally is rounded up to nearest multiple of the alignment.
+  uint64_t addMemoryUsage(uint64_t size);
```
@mciprian13 Quick question: You can increase memory usage this way, but it it not allocated to any object, right? And it is not possible to deallocate? So, it is very unusual in this sense. I think you should either clearly document this in the comment or even better make it possible to deallocate this memory somehow.

Another question: Do we need the opposite of this function so that we have incMemoryUsage(uint64_t)/decMemoryUsage(uint64_t)? if so, decMemoryUsage should probably check if any allocated segments are affected by this and then fail or free them.

mciprian13 commented:
@opti-mix The trick you proposed works but with major difference:
- The activations memory between different functions is not reused (this is why the unit tests from RELEASE_WITH_EXPENSIVE_TESTS fails because previously two functions using 64 and 128 bytes of activations respectively were allocated together 128 bytes whereas now we allocate for them 192 bytes. Is this the intended behavior?
- I think I have a fix for this:
  - I will remove the `addMemoryUsage` and use instead the function `allocate` to allocate a pool of size `activationsSize` obtained using the `allocateAll` functions and then deallocate it immediately. It is like using `allocate` to allocate a contiguous segment of activations for a given function while using `allocateAll` to allocate all the individual buffers within that segment. WDYT?

opti-mix commented:
@mciprian13 Yes, your proposal to use allocate instead of addMemoryUsage seems reasonable to me.

opti-mix commented:
@mciprian13 

We still see some assertions on CI

```
 1/55 Test  #2: BundleSaverDynamicAPITest .............Child aborted***Exception:   0.59 sec
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from BundleSaver
[ RUN      ] BundleSaver.testAPI
BundleSaverTest: ../lib/CodeGen/MemoryAllocator.cpp:202: void glow::MemoryAllocator::setHandle(uint64_t, uint64_t, glow::MemoryAllocator::Handle): Assertion `contains(ptr) && "The address is not allocated"' failed.
 #0 0x0000000002695f9f llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/var/lib/jenkins/workspace/build/tests/BundleSaverTest+0x2695f9f)
 #1 0x00000000026943c2 llvm::sys::RunSignalHandlers() (/var/lib/jenkins/workspace/build/tests/BundleSaverTest+0x26943c2)
 #2 0x0000000002696678 SignalHandler(int) (/var/lib/jenkins/workspace/build/tests/BundleSaverTest+0x2696678)
 #3 0x00007f5f82c22390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
 #4 0x00007f5f81fdb428 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x35428)
 #5 0x00007f5f81fdd02a abort (/lib/x86_64-linux-gnu/libc.so.6+0x3702a)
 #6 0x00007f5f81fd3bd7 (/lib/x86_64-linux-gnu/libc.so.6+0x2dbd7)
 #7 0x00007f5f81fd3c82 (/lib/x86_64-linux-gnu/libc.so.6+0x2dc82)
 #8 0x00000000023288ed glow::MemoryAllocator::setHandle(unsigned long, unsigned long, void const*) /var/lib/jenkins/workspace/build/../lib/CodeGen/MemoryAllocator.cpp:203:3
 #9 0x000000000232887b glow::MemoryAllocator::allocate(unsigned long, void const*) /var/lib/jenkins/workspace/build/../lib/CodeGen/MemoryAllocator.cpp:85:10
#10 0x000000000230c2f1 glow::allocateActivations(glow::TaggedList<glow::Instruction, glow::InstructionTraits> const&, glow::MemoryAllocator&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, glow::runtime::RuntimeSymbolInfo, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, glow::runtime::RuntimeSymbolInfo> > >&) /var/lib/jenkins/workspace/build/../lib/Backend/BackendUtils.cpp:463:44
```

mciprian13 commented:
@opti-mix Yeah I already figured it out. It's the corner case when there`s nothing to allocate. I`ll fix this shorty.

mciprian13 commented:
@opti-mix It seems a couple of unit tests are still failing for OpenCL.
Question: The TensorView instructions are sourced only by `AllocActivations` or can be sourced also by Constant or Mutable variables. If so then I can't use the early return as I have used it.

opti-mix commented:
@mciprian13 

> Question: The TensorView instructions are sourced only by AllocActivations or can be sourced also by Constant or Mutable variables. If so then I can't use the early return as I have used it.

Yes, TensorViews may refer to Constant or Mutable variables.

mciprian13 commented:
@opti-mix Ok. This means that it would have been better to have a separate function altogether for allocating TensorViews if they are not specifically tied to Activations. I'll make it work somehow.

mciprian13 commented:
@opti-mix There are two more things I want to do:
1. I realized that when I'm using the `allocateAll` I actually register within the allocator NOT the original sizes which were requested but rather the aligned sizes. This is harmless for the current way we're using it because after allocating we are using just the start address of the segments but in order to align it with what `allocate` does I should register the original requested sizes of the segments. This matters when someone else will use the function and might be confused (e.g. for a requested segment of size 13 with allocation of alignment 64 I actually register the segment as having size 64). This was the intention from the beginning but it makes more sense to align with the `allocate`. I will also have to change some unit tests.
2. I would like the allocations functions to be capable to allocate a segment of size 0 (corner case). If you remember some while ago when I added a mechanism at IRGen for scratch allocation, when a scratch of size 0 was requested by the node I actually allocated a scratch with 1 byte to not upset the rest of the compilation. This is wasteful because requesting 1 byte is same as requesting 64 bytes (which is the default alignment of the memory allocators). WDYT? Should I make this feature in another PR?

opti-mix commented:
@mciprian13 (1) and (2) sounds OK. (2) could be a separate PR unless it is really small.

opti-mix commented:
@mciprian13 The PR is almost good to go! Please let me know when you are done with the final changes, so that I can do a final review and we can land it.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -27,11 +27,44 @@
+  Handle handle_;
+
+  /// Allocation request type flag: true for ALLOC, false for FREE.
+  bool alloc_;
```
It could be better to put small data members like this after all other  data members due to alignment. The whole struct will use less space.

mciprian13 replied:
Done.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -27,11 +27,44 @@
+
+  /// Allocation handle which uniquely identifies the buffer to be allocated.
+  /// The provided handle must be unique for each buffer.
+  Handle handle_;
```
Glow coding style is to use underscores at the end for fields in classes. But no underscores for fields in structs.

mciprian13 replied:
Done.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -27,11 +27,44 @@
+struct Segment {
+
   /// The allocation starts at this address.
   uint64_t begin_;
```
See my comment about the naming convention.

mciprian13 replied:
Done.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -105,6 +158,9 @@ class MemoryAllocator {
   bool hasHandle(uint64_t ptr) const;
 
+  /// \returns the segment currently associated with the \p handle.
+  Segment getSegment(Handle handle) const;
```
Nit: just checking if it should be Segment or rather `const Segment &`?

mciprian13 replied:
Changed to `const Segment &`.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -156,8 +241,27 @@ class MemoryAllocator {
   /// Associates a \p handle with an allocated address \p ptr and size \p size.
   void setHandle(uint64_t ptr, uint64_t size, Handle handle);
+
+  /// Function to verify the allocations before allocating the segments with
```
Could you describe the parameters of the function in the comment?

mciprian13 replied:
Done.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -156,8 +241,27 @@ class MemoryAllocator {
+  /// otherwise.
+  bool verifyAllocations(const std::list<Allocation> &allocList) const;
+
+  /// Function to verify the segments allocated with \ref allocateAll.
```
Could you describe the parameters of the function in the comment?

mciprian13 replied:
Done.

opti-mix commented on include/glow/CodeGen/MemoryAllocator.h:
```diff
@@ -156,8 +241,27 @@ class MemoryAllocator {
+  /// \returns true if segments are valid and false otherwise.
+  bool verifySegments(const std::list<Allocation> &allocList) const;
+
+  /// Utility function to map the handles used in the allocations \p allocList
```
Could you describe the parameters of the function in the comment?

mciprian13 replied:
This was already pretty explicit.

opti-mix commented on lib/Backend/BackendUtils.cpp:
```diff
@@ -34,7 +34,7 @@ static llvm::cl::OptionCategory BackendUtilsCat("Glow Backend Utils Options");
 static llvm::cl::opt<bool> reuseActivationsMemory(
     "reuse-activation-memory-allocations",
-    llvm::cl::desc("Should activation memory allocations be reused"),
+    llvm::cl::desc("Reuse activations memory between different IR functions"),
```
Here you repurpose the flag for something different. 

The original intent of this flag was to prohibit reuse (at the expense of using more memory), so that the memory for each activation is never used for anything else. It is useful for debugging on e.g. new HW/bare metal, where you want to check post-mortem, after you dumped all of the memory used jointly by activations, that memory segments for specific activations have specific values.

mciprian13 replied:
The meaning of this flag has indeed changed from its previous use (cannot do otherwise). Previously this flag enabled us to allocate exclusive activations buffers within functions and between functions. Now it can only enable us to allocate exclusive activations between functions but not within functions (the buffers within same function will always be allocated in a compact-optimized way). But you said that you don't mind if I break it. Instead I repurposed it for something similar.

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  return true;
+}
+
+/// Buffer information structure.
```
Maybe move the type definitions to the beginning of the file?

mciprian13 replied:
I'm not sure I want to move it since this is only related to `allocateAll`. I want all the code related to `allocateAll` to be centralized for easier navigation.

mciprian13 commented:
@opti-mix I managed to do 1) and added extra unit tests. For 2) I added the capability to allocate segments of size 0 only for the function `allocateAll` (plus unit tests). For the function `allocate` if I try to do that I open Pandora's box because the function `allocate` uses a `addrToHandleMap` which in my view is not only redundant but screws up things when using segments of size 0.
Please let me know when you finish the review so that I can address all the items at once.

opti-mix commented on tests/unittests/MemoryAllocatorTestModels.txt:
```diff
@@ -0,0 +1,8254 @@
+MODEL cifar10_quant.tflite
```
It would be nice to describe somewhere in the docs or source comments how this file was produced and how to update it if e.g. a model changes or a new model needs to be added.

mciprian13 replied:
Added a header within this file about how it was obtained and how somebody can add manually another model:
```
################################################################################
# This file describes a sequence of activation memory allocation/deallocation
# instructions for multiple models. This file is used to test the function
# "allocateAll()" from the MemoryAllocator object. This file was generated
# automatically using a Matlab script which parses the Glow IR obtained during
# Glow compilation with the option -dump-ir for all the models. This text file
# has a <key, value(s)> format with the following convention:
#   NAME              - Model name (used as delimiter between models).
#   ALIGN <ALIGNMENT> - Alignment (in bytes) used for memory allocator.
#   ALLOC <ID> <SIZE> - Instruction to allocate a buffer with identifier <ID>
#                       and size <SIZE>. Each buffer must have a unique ID.
#   FREE <ID>         - Instruction to free a previously allocated buffer with
#                       identifier <ID>. For a given model every buffer which
#                       is allocated must be later freed.
#   MEM <SIZE>        - Expected memory usage for the alloc/free sequence.
#   EFF <RATIO>       - Expected efficiency of the allocation
# You can define manually additional unit tests following the given convention.
################################################################################
```

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+static std::vector<MemAllocStrategy> memAllocStrategies = {
+    MaxLiveSizeMaxBuffSize, MaxLiveSizeMaxBuffTime, SameOrder};
+
+/// Utility function to allocate all the segments at once using the given
```
Please describe all parameters here and in other functions.

mciprian13 replied:
Done.

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  for (const auto &alloc : allocList) {
+    // We only map the Handles of ALLOCs.
+    if (alloc.alloc_) {
+      handleToIdMap[alloc.handle_] = id;
```
Nit: It seems like it assumes that each alloc has a unique handle. I'm not sure it is really guaranteed, though it is probably the case today. Strictly speaking, if there are two allocations with non-overlapping life times, they could (re)-use the same handle.

mciprian13 replied:
From `allocateAll` point of view all the handles must be unique in order to later identify the allocated segments even if they are non-overlapping (otherwise the whole segment to handle map is ambiguous).

opti-mix replied:
OK. But this is an important requirement. It should be explicitly explained in the comments for the memory allocator or methods related to handles.

mciprian13 replied:
It is stated in the Allocation structure that each handle must be unique.

opti-mix commented on lib/CodeGen/MemoryAllocator.cpp:
```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  assert(id == buffNum && "Inconsistent Handle to ID mapping!");
+}
+
+uint64_t MemoryAllocator::allocateAll(const std::list<Allocation> &allocList,
```
Nit: This function is still pretty big. Don't know if it is possible to split it into smaller parts.

mciprian13 replied:
Yeah I`m not sure I can split it further.

opti-mix commented:
@mciprian13 I think I'm done with my review mostly. Once you address the issues, I'll import it into our internal build system and see if it breaks anything. if everything is green, we could land this PR.

mciprian13 commented:
@opti-mix Please have another look.

opti-mix commented:
@mciprian13 Looking good! Let me try to import it into our internal CI and see if there are any issues.

opti-mix commented:
@mciprian13 This diff results in breaking one of the closed-sourced end-2-end models runs on our side. We are investigating what is the issue.

mciprian13 commented:
@opti-mix Thanks for your effort in getting this PR landed. I really appreciate it.

opti-mix commented:
@mciprian13 

> This diff results in breaking one of the closed-sourced end-2-end models runs on our side. We are investigating what is the issue.

Update: Your PR exposed a very subtle bug in one of our internal models. It took some time to figure out the root cause. We are working on a fix on our side. Hopefully it can be fixed in a couple of days or next week.

opti-mix commented on lib/Backend/BackendUtils.cpp:
```diff
@@ -433,10 +433,47 @@ void allocatePlaceholders(const ContiguousPlaceholders &placeholders,
                          MemoryAllocator &allocator,
                          glow::runtime::SymbolTableTy &symbolTable) {
+
+  // Gather allocation/deallocation sequence.
```
@mciprian13 

While debugging the issue we found on our side, I found it useful to add a command-line option telling the memory allocation logic to never reuse memory for activations (so that you can easily inspect the value of each activation and to make sure that two activations never use the same memory).

Achieving this is pretty easy: Simply insert all the deallocations at the very end of the allocList. This effectively makes all activations alive for the whole duration of the IRFunction and thus they never reuse memory.

If it is not too much trouble, I'd suggest you add such a command-line debug option and the corresponding logic here. Should be just about 10-15 lines of code I think.

mciprian13 replied:
@opti-mix So you mean to restore the old functionality of the `reuseActivationsMemory`? Just to confirm what you want when this flag is false is to have non-overlapping activations both within same function and also between different functions, right? As you said that is easily doable.

opti-mix replied:
Yes, I mean exactly this. Thanks for implementing it.
